From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ethan Chan <ethanchan7981@gmail.com>
Date: Wed, 21 Aug 2024 01:33:11 -0400
Subject: [PATCH] Use Platform Math Functions


diff --git a/src/main/java/net/minecraft/util/Mth.java b/src/main/java/net/minecraft/util/Mth.java
index f298cdfcf1539e467f57f9f7789de3cf2ca54665..a672c8aae5cf20a305956441f33d8a1b6981e0e0 100644
--- a/src/main/java/net/minecraft/util/Mth.java
+++ b/src/main/java/net/minecraft/util/Mth.java
@@ -58,13 +58,11 @@ public class Mth {
     }
 
     public static int floor(float value) {
-        int i = (int)value;
-        return value < (float)i ? i - 1 : i;
+        return (int) Math.floor(value);
     }
 
     public static int floor(double value) {
-        int i = (int)value;
-        return value < (double)i ? i - 1 : i;
+        return (int) Math.floor(value);
     }
 
     public static long lfloor(double value) {
@@ -81,13 +79,11 @@ public class Mth {
     }
 
     public static int ceil(float value) {
-        int i = (int)value;
-        return value > (float)i ? i + 1 : i;
+        return (int) Math.ceil(value);
     }
 
     public static int ceil(double value) {
-        int i = (int)value;
-        return value > (double)i ? i + 1 : i;
+        return (int) Math.ceil(value);
     }
 
     public static int clamp(int value, int min, int max) {
@@ -123,15 +119,7 @@ public class Mth {
     }
 
     public static double absMax(double a, double b) {
-        if (a < 0.0) {
-            a = -a;
-        }
-
-        if (b < 0.0) {
-            b = -b;
-        }
-
-        return Math.max(a, b);
+        return Math.max(Math.abs(a), Math.abs(b));
     }
 
     public static int floorDiv(int dividend, int divisor) {
